janitor/functions/collapse_levels.py

```diff
@@ -78,11 +79,32 @@ class  max_speed    type
     if not isinstance(df.columns, pd.MultiIndex):
         return df
 
-    df = df[:]
+    # TODO: Pyarrow offers faster string computations
+    # future work should take this into consideration,
+    # which would require a different route from python's string.join
 
-    df.columns = [
-        sep.join(str(el) for el in tup if str(el) != "")
-        for tup in df  # noqa: PD011
+    # since work is only on the columns
+    # it is safe, and more efficient to slice/view the dataframe
+    # plus Pandas creates a new Index altogether
+    # as such, the original dataframe is not modified
+    df = df[:]
+    new_columns = df.columns
+    levels = [
+        new_columns.get_level_values(num) for num in range(new_columns.nlevels)
     ]
-
+    all_strings = all(map(is_string_dtype, levels))
+    if all_strings:
+        no_empty_string = all((entry != "").all() for entry in levels)
+        if no_empty_string:
+            df.columns = new_columns.map(sep.join)
+            return df
+    new_columns = (map(str, entry) for entry in new_columns)
+    new_columns = [
+        # faster to use a list comprehension within string.join
+        # compared to a generator
+        # https://stackoverflow.com/a/37782238
```

Comment:
Lovely comment here! The rationale and source all cited for future devs to see. Thank you for doing it the awesome way, @samukweku!